cli/flox-activations/src/cli/start_or_attach.rs

```diff
@@ -0,0 +1,128 @@
+use std::path::{Path, PathBuf};
+
+use anyhow::Context;
+use clap::Args;
+use indoc::indoc;
+use time::{Duration, OffsetDateTime};
+
+use crate::activations;
+
+#[derive(Debug, Args)]
+pub struct StartOrAttachArgs {
+    #[arg(help = "The PID of the shell registering interest in the activation.")]
+    #[arg(short, long, value_name = "PID")]
+    pub pid: u32,
+    #[arg(help = "The path to the .flox directory for the environment.")]
+    #[arg(short, long, value_name = "PATH")]
+    pub flox_env: PathBuf,
+    #[arg(help = "The store path of the rendered environment for this activation.")]
+    #[arg(short, long, value_name = "PATH")]
+    pub store_path: String,
+}
+
+impl StartOrAttachArgs {
+    pub(crate) fn handle(self, cache_dir: PathBuf) -> Result<(), anyhow::Error> {
+        let activations_json_path = activations::activations_json_path(&cache_dir, &self.flox_env)?;
+
+        let (activations, lock) = activations::read_activations_json(&activations_json_path)?;
+        let mut activations = activations.unwrap_or_default();
+
+        let (activation_id, attaching) = match activations
+            .activation_for_store_path(&self.store_path)
+        {
+            Some(activation) => {
+                let activation_id = activation.id();
+
+                let attach_expiration = OffsetDateTime::now_utc() + Duration::seconds(10);
+
+                // While we wait for the activation to become ready,
+                // we need to drop the lock to allow other processes to write to the file.
+                drop(lock);
+                wait_for_activation_ready_and_attach_pid(
+                    &activations_json_path,
+                    activation_id,
+                    attach_expiration,
+                    self.pid,
+                )?;
+
+                (activation_id, true)
+            },
+            None => {
+                let activation_id = activations
+                    .create_activation(&self.store_path, self.pid)?
+                    .id();
+
+                activations::write_activations_json(&activations, &activations_json_path, lock)?;
+
+                (activation_id, false)
+            },
+        };
+
+        println!("_FLOX_ATTACH={attaching}");
+        println!(
+            "_FLOX_ACTIVATION_STATE_DIR={}",
+            activations::activation_state_dir_path(cache_dir, self.flox_env, activation_id)?
+                .display()
+        );
+        println!("_FLOX_ACTIVATION_ID={activation_id}");
+
+        Ok(())
+    }
+}
+
+/// Wait for the activation with the given ID to become ready.
+/// I.e if an activation is being started already, wait for it to become ready,
+/// then _attach_ the given PID to it.
+/// If the activation is not ready within the given timeout,
+/// exit with an error.
+/// If the activation startup process fails, exit with an error.
+/// In either case, the activation can likely just be restarted.
+fn wait_for_activation_ready_and_attach_pid(
```

Comment:
```suggestion
// TODO: use store_path rather than activation_id.
// If we are activation 3, activation 1 fails, but activation 2 succeeds, we may want to attach to activation 2
fn wait_for_activation_ready_and_attach_pid(
```